docs/src/literate-tutorials/linear_elasticity.jl

```diff
@@ -1,3 +1,243 @@
+ch = ConstraintHandler(dh)
+add!(ch, Dirichlet(:u, getfaceset(grid, "bottom"), (x, t) -> 0.0, 2))
+add!(ch, Dirichlet(:u, getfaceset(grid, "left"), (x, t) -> 0.0, 1))
+add!(ch, Dirichlet(:u, getfaceset(grid, "top"), (x, t) -> 0.1, 2))
+close!(ch);
+
+# ### Material routine
+# For the sake of structuring the program, we introduce a material routine here.
+# We define a data structure that holds the material constants, here it stores the
+# shear modulus $G$ and the bulk modulus $K$. The material routine then computes the stress
+# $\boldsymbol{\sigma}$ and its tangent $\frac{\partial\boldsymbol{\sigma}}{\partial\boldsymbol{\varepsilon}}$
+# based on the input strain $\boldsymbol{\varepsilon}$. Here, we use automatic differentiation
+# for computing the stress tangent. You can read more about automatic differentiation on tensor
+# operations in the [`Tensors.jl` docs](https://ferrite-fem.github.io/Tensors.jl/stable/man/automatic_differentiation/).
+struct Elasticity
+    G::Float64
+    K::Float64
+end
+
+function material_routine(material::Elasticity, ε::SymmetricTensor{2})
+    (; G, K) = material
+    stress(ε) = 2G * dev(ε) + K * tr(ε) * one(ε)
+    ∂σ∂ε, σ = gradient(stress, ε, :all)
+    return σ, ∂σ∂ε
+end
+
+E = 200e3 # Young's modulus [MPa]
+ν = 0.3 # Poisson's ratio [-]
+material = Elasticity(E/2(1+ν), E/3(1-2ν));
+
+# ### Element routine
+# The residual vector and stiffness matrix follow from the weak form such that
+# ```math
+# \left(\underline{R}\right)_i
+# =
+# \int_\Omega
+#   \boldsymbol{\sigma} : \boldsymbol{\nabla} \boldsymbol{N}_i
+# \, \mathrm{d}V
+# \,, \qquad
+# \left(\underline{\underline{K}}\right)_{ij}
+# =
+# \int_\Omega
+#   \left(
+#       \frac{\partial \boldsymbol{\sigma}}{\partial \boldsymbol{\varepsilon}}
+#       :
+#       \boldsymbol{\nabla}^\mathrm{sym} \boldsymbol{N}_j
+#   \right)
+#   :
+#   \boldsymbol{\nabla} \boldsymbol{N}_i
+# \, \mathrm{d}V
+# ```
+# The element routine computes the local residual vector `re` and stiffness matrix `ke`
+# for a single element. `ke` and `re` are pre-allocated and reused for all elements.
+function assemble_cell!(ke, fe, cellvalues, material, ue)
+    fill!(ke, 0.0)
+    fill!(fe, 0.0)
+
+    n_basefuncs = getnbasefunctions(cellvalues)
+    for q_point in 1:getnquadpoints(cellvalues)
+        ## For each integration point, compute strain, stress and material stiffness
+        ε = function_symmetric_gradient(cellvalues, q_point, ue)
+        σ, ∂σ∂ε = material_routine(material, ε)
+
+        dΩ = getdetJdV(cellvalues, q_point)
+        for i in 1:n_basefuncs
+            ∇Nᵢ = shape_gradient(cellvalues, q_point, i)# shape_symmetric_gradient(cellvalues, q_point, i)
+            fe[i] += σ ⊡ ∇Nᵢ * dΩ # add internal force to residual
+            for j in 1:n_basefuncs
+                ∇ˢʸᵐNⱼ = shape_symmetric_gradient(cellvalues, q_point, j)
+                ke[i, j] += (∂σ∂ε ⊡ ∇ˢʸᵐNⱼ) ⊡ ∇Nᵢ * dΩ
+            end
+        end
+    end
+end
+#md nothing # hide
```

Comment:
So you implement the solver via one Newton step if I understand this correctly? Maybe we should comment in this case.

Reply:
I usually do that, so this is sort of an artifact. Removing it now for making this truely linear.